In [36]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn.feature_selection import SelectKBest, f_regression

# Загрузка данных

In [37]:
%%capture
!wget https://www.dropbox.com/s/64ol9q9ssggz6f1/data_ford_price.xlsx

In [38]:
data = pd.read_excel('C:/Users/админ/Desktop/data_ford_price.xlsx')

# Предобработка данных

In [39]:
data = data[['price','year', 'cylinders', 'odometer', 'lat', 'long', 'weather']]
data.dropna(inplace = True)

y = data['price']
x = data.drop(columns='price')

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

# 1. Выбор 3-х признаков по методу RFE и обучение модели на отобранных признаках

In [40]:
#Выделим три наиболее значимых признака при помощи метода RFE:
from sklearn.feature_selection import RFE
 
estimator = LinearRegression()
selector = RFE(estimator, n_features_to_select=3, step=1)
selector = selector.fit(X_train, y_train)
 
selector.get_feature_names_out()

array(['year', 'cylinders', 'lat'], dtype=object)

In [41]:
#Предобработк данных
data_copy=data.copy()
data_copy = data_copy[['price','year', 'cylinders',  'lat', ]]
data_copy.dropna(inplace = True)

y = data_copy['price']
x = data_copy.drop(columns='price')

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

#Обучаем модель
model = LinearRegression()
model.fit(X_train, y_train)
y_predicted = model.predict(X_test)
 
mae = mean_absolute_error(y_test, y_predicted)
print('MAE: %.3f' % mae)

MAE: 5096.570


# 2. Выбор 3-х признаков по методу SelectKBest и обучение модели на отобранных признаках

In [43]:
data = pd.read_excel('C:/Users/админ/Desktop/data_ford_price.xlsx')

In [44]:
data = data[['price','year', 'cylinders', 'odometer', 'lat', 'long', 'weather']]
data.dropna(inplace = True)

y = data['price']
x = data.drop(columns='price')

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

In [45]:
#Выделим три наиболее значимых признака при помощи метода SelectKBest:
from sklearn.feature_selection import SelectKBest, f_regression
selector_KBest = SelectKBest(f_regression, k=3)
selector_KBest.fit(X_train, y_train)
 
selector_KBest.get_feature_names_out()

array(['year', 'cylinders', 'odometer'], dtype=object)

In [46]:
#Предобработк данных
data_copy1=data.copy()
data_copy1 = data_copy1[['price','year', 'cylinders',  'odometer', ]]
data_copy1.dropna(inplace = True)

y = data_copy1['price']
x = data_copy1.drop(columns='price')

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

#Обучаем модель
model = LinearRegression()
model.fit(X_train, y_train)
y_predicted = model.predict(X_test)
 
mae = mean_absolute_error(y_test, y_predicted)
print('MAE: %.3f' % mae)

MAE: 4708.946


# Вывод:
Средняя абсолютная ошибка модели регрессии, обученной на первых трех столбцах ( былои выбраны по методу RFE) составила 5096.570, а средняя абсолютная ошибка модели регрессии, обученной на вторых трех столбцах ( былои выбраны по методу SelectKBest) составила 4708.946. Отсюда делаем вывод, что  метод отбора  признаков SelectKBest показал наилучший результат на тестовой выборке, так как средняя абсолютная ошибка меньше, чем при методе отбора признаков RFE.